rust/src/ldap/ldap.rs

```diff
@@ -0,0 +1,426 @@
+
+use crate::applayer::{self, *};
+use crate::conf::conf_get;
+use crate::core::{AppProto, Flow, ALPROTO_UNKNOWN, IPPROTO_TCP};
+use ldap_parser::ldap::LdapMessage as LdapParserMessage;
+use ldap_parser::FromBer;
+use nom7 as nom;
+use std;
+use std::collections::VecDeque;
+use std::ffi::CString;
+use std::os::raw::{c_char, c_int, c_void};
+
+use crate::ldap::types::*;
+
+static mut LDAP_MAX_TX: usize = 256;
+
+static mut ALPROTO_LDAP: AppProto = ALPROTO_UNKNOWN;
+
+#[derive(AppLayerEvent)]
+enum LdapEvent {
+    TooManyTransactions,
+}
+
+#[derive(Debug)]
+pub struct LdapTransaction {
+    pub tx_id: u64,
+    pub request: Option<LdapMessage>,
+    pub responses: VecDeque<LdapMessage>,
+    complete: bool,
+
+    tx_data: AppLayerTxData,
+}
+
+impl Default for LdapTransaction {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl LdapTransaction {
+    pub fn new() -> LdapTransaction {
+        Self {
+            tx_id: 0,
+            request: None,
+            responses: VecDeque::new(),
+            complete: false,
+            tx_data: AppLayerTxData::new(),
+        }
+    }
+
+    fn set_event(&mut self, e: LdapEvent) {
+        self.tx_data.set_event(e as u8);
+    }
+}
+
+impl Transaction for LdapTransaction {
+    fn id(&self) -> u64 {
+        self.tx_id
+    }
+}
+
+#[derive(Default)]
+pub struct LdapState {
+    state_data: AppLayerStateData,
+    tx_id: u64,
+    transactions: VecDeque<LdapTransaction>,
+    request_gap: bool,
+    response_gap: bool,
+    tx_index_completed: usize,
+}
+
+impl State<LdapTransaction> for LdapState {
+    fn get_transaction_count(&self) -> usize {
+        self.transactions.len()
+    }
+
+    fn get_transaction_by_index(&self, index: usize) -> Option<&LdapTransaction> {
+        self.transactions.get(index)
+    }
+}
+
+impl LdapState {
+    pub fn new() -> Self {
+        Self {
+            state_data: AppLayerStateData::new(),
+            tx_id: 0,
+            transactions: VecDeque::new(),
+            request_gap: false,
+            response_gap: false,
+            tx_index_completed: 0,
+        }
+    }
+
+    // Free a transaction by ID.
+    fn free_tx(&mut self, tx_id: u64) {
+        let len = self.transactions.len();
+        let mut found = false;
+        let mut index = 0;
+        for i in 0..len {
+            let tx = &self.transactions[i];
+            if tx.tx_id == tx_id + 1 {
+                found = true;
+                index = i;
+                break;
+            }
+        }
+        if found {
+            self.transactions.remove(index);
+        }
+    }
+
+    pub fn get_tx(&mut self, tx_id: u64) -> Option<&LdapTransaction> {
+        self.transactions.iter().find(|tx| tx.tx_id == tx_id + 1)
+    }
+
+    pub fn new_tx(&mut self) -> LdapTransaction {
+        let mut tx = LdapTransaction::new();
+        self.tx_id += 1;
+        tx.tx_id = self.tx_id;
+        if self.transactions.len() > unsafe { LDAP_MAX_TX } {
```

Comment:
Wonder what the best way is here, if we accept gaps.

Like I have request 1, request 2, gap, request 4
And I have response1, gap, response 3, response 4

Should I drop request 2 because I expect that response 2 was in the gap ?